netty/src/main/java/io/grpc/transport/netty/NettyClientTransport.java

```diff
@@ -182,7 +183,13 @@ public ClientStream newStream(MethodDescriptor<?, ?> method, Metadata.Headers he
     Http2Headers http2Headers = Utils.convertClientHeaders(headers, ssl, defaultPath, authority);
 
     // Write the command requesting the creation of the stream.
-    channel.writeAndFlush(new CreateStreamCommand(http2Headers, stream));
+    CreateStreamCommand msg = new CreateStreamCommand(http2Headers, stream);
+    if (method.getType().clientSendsOneMessage()) {
+      // No need to flush yet if the client is required to send at least one message
+      channel.write(msg);
```

Comment:
@nmittler, I share your concern, but I don't agree with your alternative suggestion. It is true that the only case to improve is when MAX_CONCURRENT_STREAMS is exceeded, though.

Tagging `CreateStreamCommand` alone won't solve the flushing problem because `BufferingHttp2ConnectionEncoder` is _after_ the `CreateStreamCommand` processing; our handler won't know that the stream will be buffered when processing `CreateStreamCommand`. We would need to add a method to `BufferingHttp2ConnectionEncoder` to inform it of the conditional flushes.

I actually plan a different route for fully-optimized flushing: buffer writes per stream on the application thread until the flush, at which time we do a single execute in the channel's event loop. This has lower synchronization overhead, uses marginally less memory, and improves the flush behavior when MAX_CONCURRENT_STREAMS is exceeded.